### PR TITLE
Refactor model runners

### PR DESCRIPTION
**Pull request by mitchellgoffpc**

### CODE REVIEW

File: selfdrive/modeld/models/driving.cc [hunk 1]
```diff
@@ -33,26 +33,30 @@ void model_init(ModelState* s, cl_device_id device_id, cl_context context) {
...
-  s->m->addDrivingStyle(s->driving_style, DRIVING_STYLE_LEN);
+  s->m->addInput("driving_style", s->driving_style, DRIVING_STYLE_LEN);
 #endif
 
 #ifdef NAV
-  s->m->addNavFeatures(s->nav_features, NAV_FEATURE_LEN);
+  s->m->addInput("nav_features", s->nav_features, NAV_FEATURE_LEN);
+#endif
+
+#ifdef TEMPORAL
```
Comment (adeebshihadeh) [inline selfdrive/modeld/models/driving.cc:58]: do we ever not set these macros? think it's some legacy stuff that's been carried over forever

File: selfdrive/modeld/runners/runmodel.h [hunk 2]
```diff
@@ -1,18 +1,19 @@
 #pragma once
 #include "common/clutil.h"
+
+struct ModelInput {
+  const char* name;
+  float *buffer;
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/runmodel.h:6]: ptr style isn't consistent

File: selfdrive/modeld/runners/onnxmodel.h [hunk 3]
```diff
@@ -1,51 +1,28 @@
...
-  void addImage(float *image_buf, int buf_size);
-  void addExtra(float *image_buf, int buf_size);
+  void addInput(const char *name, float *buffer, int size);
+  void setInputBuffer(const char *name, float *buffer, int size);
   void execute();
 private:
   int proc_pid;
 
+  std::vector<ModelInput> inputs;
   float *output;
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/onnxmodel.h:19]: move to RunModel?

File: selfdrive/modeld/runners/onnxmodel.cc [hunk 4]
```diff
@@ -87,72 +86,25 @@ void ONNXModel::pread(float *buf, int size) {
...
-
-void ONNXModel::addImage(float *image_buf, int buf_size) {
-  image_input_buf = image_buf;
-  image_buf_size = buf_size;
-}
-
-void ONNXModel::addExtra(float *image_buf, int buf_size) {
-  extra_input_buf = image_buf;
-  extra_buf_size = buf_size;
+void ONNXModel::setInputBuffer(const char *name, float *buffer, int size) {
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/onnxmodel.cc:93]: the setInputBuffer code is mostly the same. can you pull those parts out into a common one?

File: selfdrive/modeld/runners/onnxmodel.cc [hunk 5]
```diff
@@ -87,72 +86,25 @@ void ONNXModel::pread(float *buf, int size) {
...
+  for (auto &input : inputs) {
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/onnxmodel.cc:94]: replace with std::map?

File: selfdrive/modeld/runners/thneedmodel.cc [hunk 6]
```diff
@@ -1,78 +1,75 @@
...
+    }
+  }
 
-void* ThneedModel::getInputBuf() {
-  if (use_extra && thneed->input_clmem.size() > 5) return &(thneed->input_clmem[5]);
-  else if (!use_extra && thneed->input_clmem.size() > 4) return &(thneed->input_clmem[4]);
-  else return nullptr;
-}
+  if (index == -1) {
+    LOGE("Tried to get CL buffer for input `%s` but no input with this name exists", name);
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/thneedmodel.cc:43]: assert?

File: selfdrive/modeld/models/driving.cc (hunk 1 above)
Comment (mitchellgoffpc) [reply selfdrive/modeld/models/driving.cc:58]: not that I'm aware of, but I think removing these should be a separate PR

File: selfdrive/modeld/runners/runmodel.h [hunk 7]
```diff
@@ -1,18 +1,43 @@
 #pragma once
+#include <string>
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/runmodel.h:2]: add a new line above this

File: selfdrive/modeld/runners/runmodel.h [hunk 8]
```diff
@@ -1,18 +1,43 @@
...
+#include <vector>
+#include <memory>
+#include <cassert>
 #include "common/clutil.h"
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/runmodel.h:6]: and one above here

**Pull request merged**